client/mysqltest.cc

```diff
@@ -3865,26 +3865,268 @@ void do_remove_files_wildcard(struct st_command *command)
 }
 
 
+/*
+  SYNOPSIS
+  recursive_copy
+  ds_source      - pointer to dynamic string containing source
+                   directory information
+  ds_destination - pointer to dynamic string containing destination
+                   directory information
+  DESCRIPTION
+    Recursive copy of <ds_source> to <ds_destination>
+
+    For simplicity, always create directory <ds_destination>/<source_dir_name>
+    and then copy the content to it. For example `copy_file a/b $dst -r` will
+    always create directory $dst/b
+
+    Note: When symlink appears, copy the target file
+*/
+
+static int recursive_copy(DYNAMIC_STRING *ds_source,
+                          DYNAMIC_STRING *ds_destination,
+                          int do_not_overwrite,
+                          bool create_new_dir_in_dst)
+{
+  /* Note that my_dir sorts the list if not given any flags */
+  MY_DIR *src_dir_info= my_dir(ds_source->str,
+                               MYF(MY_DONT_SORT | MY_WANT_STAT));
+
+  int error= 0;
+
+  /* Source directory exists */
+  if (src_dir_info)
+  {
+    /* Note that my_dir sorts the list if not given any flags */
+    MY_DIR *dest_dir_info= my_dir(ds_destination->str,
+                                  MYF(MY_DONT_SORT | MY_WANT_STAT));
+
+    /* Create destination directory if it doesn't exist */
+    if (!dest_dir_info)
+    {
+      error= my_mkdir(ds_destination->str, 0777, MYF(0)) != 0;
+      if (error)
+      {
+        my_dirend(dest_dir_info);
+        goto end;
+      }
+    }
+
+    /* Extracting the source basename */
+    if (ds_source->str[strlen(ds_source->str) - 1] == '/')
+    {
+      ds_source->str[strlen(ds_source->str) - 1]= '\0';
+      ds_source->length= ds_source->length - 1;
+    }
+    char *src_dir_name= strrchr(ds_source->str, '/');
+
+    /* Create <destination_directory>/<source_dir_name> if it doesn't exist
+       Use create_new_dir_in_dst flag because this is only and always needed
+       for the first level of recursive copy */
+    if (create_new_dir_in_dst)
+    {
+      dynstr_append_mem(ds_destination, src_dir_name, strlen(src_dir_name));
+      my_dirend(dest_dir_info);
+      dest_dir_info= my_dir(ds_destination->str,
+                            MYF(MY_DONT_SORT | MY_WANT_STAT));
+
+      if (!dest_dir_info)
+      {
+        error= my_mkdir(ds_destination->str, 0777, MYF(0)) != 0;
+        if (error)
+        {
+          my_dirend(dest_dir_info);
+          goto end;
+        }
+      }
+    }
+
+    char dir_separator[2]= {FN_LIBCHAR, 0};
```

Comment:
I'm not sure about it. Above you always compare with '/'
why here is it suddenly FN_LIBCHAR?